### ISSUE TITLE
Custom error message when authentication fails

### ISSUE THREAD

**Issue by mahendrapaipuri** [jupyterhub/jupyterhub#4421]
### Proposed change

Currently we can configure custom authenticators for JupyterHub which is quite nice. Custom authenticators can have a varying Auth flow logic that can have multiple verifications. Normally we deny the authentication of the user if any of these steps in the auth flow fails. As of now Authenticator should return username after successful authentication or `None`. Whenever Authenticator returns `None`, frontend always displays the same message `Invalid username or password`. This might not be the entirely true for custom auth flow. An example can be the IP address of the client is not in whitelisted IPs. So, the message in the frontend can be bit misleading in these cases. 

So the proposal is to let the authenticator to return a optional message along with `None` when authentication fails. We use this message, if set, or default message, if not set to render the HTML template or default message.

### Who would use this feature?

JupyterHub deployments that use custom authenticators to implement their own auth flow. 

### (Optional): Suggest a solution

Returning an optional message from authenticator when auth fails can be a straight forward solution with backward compatibility. Error message is already configurable in Jinja template but it uses a default message now.

**Comment by minrk** [comment]
I think we can make this experience better, but you _can_ currently raise an HTTPError with a message, and it will be shown to the user:

```
from tornado import web

...
def authenticate(self, handler, data):
    raise web.HTTPError(403, "my custom message") # or 400 or 500 or whatever seems appropriate in your case
```

A difference is that this will be shown on full error page, instead of the little 'login failed' message alongside the login form.

But it would probably be a better experience to raise a special AuthenticationFailed error, where you could put a custom message, and which would be displayed more consistently with the default 'Invalid username or password' message.

**Comment by minrk** [comment]
Yes, I agree that we should be able to customize the current login error message, not just show a full error page, so I'll leave this open as a feature request.

**Comment by mahendrapaipuri** [comment]
I can work on a PR to implement this feature.
